flag/pflag.go

```diff
@@ -0,0 +1,365 @@
+func NewPFlagSetWithArgs(cfg *NameConfig, template interface{}, args []string) (*PflagSet, error) {
+	pval, ptyp, ptrifyErr := ptrified(template)
+	if ptrifyErr != nil {
+		return nil, ptrifyErr
+	}
+
+	fs := pflag.NewFlagSet("", pflag.ContinueOnError)
+
+	s := PflagSet{
+		Flags:           fs,
+		ParseFunc:       func() error { return fs.Parse(args) },
+		ptrType:         ptyp,
+		flagsRegistered: true,
+		NameCfg:         cfg,
+		flagFieldName:   map[string]string{},
+		flagValues:      map[string]interface{}{},
+	}
+
+	if err := s.registerPFlags(pval, ptyp); err != nil {
+		return nil, err
+	}
+
+	return &s, nil
+}
+
+// PflagSet is a flagset. Please only use PflagSet when using cobra command line
+// tool. Use Set for flags functionality for other cases
+type PflagSet struct {
+	Flags     *pflag.FlagSet
+	ParseFunc func() error
+
+	ptrType reflect.Type
+
+	// NameCfg defines tunables for constructing flag-names
+	NameCfg *NameConfig
+
+	flagsRegistered bool
+	tfmr            *transform.Transformer
+	trnslVal        reflect.Value
+	// Map to store the flag name (key) and field name (value)
+	flagFieldName map[string]string
+	flagValues    map[string]interface{}
+}
+
+func (s *PflagSet) parse() error {
+	if s.ParseFunc == nil {
+		return fmt.Errorf("unparsed flagset with no ParseFunc set")
+	}
+	if err := s.ParseFunc(); err != nil {
+		return fmt.Errorf("failed to parse pflags: %s", err)
+	}
+	return nil
+}
+
+func (s *PflagSet) registerPFlags(tmpl reflect.Value, ptyp reflect.Type) error {
+	fm := transform.NewFlattenMangler(transform.DialsTagName, s.NameCfg.FieldNameEncodeCasing, s.NameCfg.TagEncodeCasing)
+	tfmr := transform.NewTransformer(ptyp, fm)
+	val, TrnslErr := tfmr.Translate()
+	if TrnslErr != nil {
+		return TrnslErr
+	}
+
+	s.tfmr = tfmr
+	s.trnslVal = val
+
+	t := val.Type()
+
+	k := t.Kind()
+	for k == reflect.Ptr {
+		t = t.Elem()
+		k = t.Kind()
+	}
+
+	// the input kind will be struct after calling Translate on it
+	for i := 0; i < t.NumField(); i++ {
+		sf := t.Field(i)
+		help := DefaultFlagHelpText
+		if x, ok := sf.Tag.Lookup(HelpTextTag); ok {
+			help = x
+		}
+
+		name := s.mkname(sf)
+		s.flagFieldName[name] = sf.Name
+
+		// if the flag already exists, don't register so the user can override
+		// our behavior
+		if s.Flags.Lookup(name) != nil {
+			continue
+		}
+
+		ft := sf.Type
+
+		k := ft.Kind()
+		for k == reflect.Ptr {
+			ft = ft.Elem()
+			k = ft.Kind()
+		}
+		isValue := ft.Implements(flagReflectType) || reflect.PtrTo(ft).Implements(flagReflectType)
+		isTextM := ft.Implements(textMReflectType) || reflect.PtrTo(ft).Implements(textMReflectType)
+
+		// get the concrete value of the field from the template
+		fieldVal := transform.GetField(sf, tmpl)
+		shorthand, _ := sf.Tag.Lookup(dialsPFlagShortTag)
+		var f interface{}
+
+		switch {
+		case isValue:
+			{
+				f = s.Flags.VarPF(fieldVal.Addr().Interface().(pflag.Value), name, shorthand, help)
+				s.flagValues[name] = f
```

Comment:
ugg, it looks like the return value of `VarPF` isn't consistent with other registration methods, and returns the flag instead of a pointer to the value. :facepalm: 
I think we want to assign `fieldVal.Addr().Interface()` into the `flagValues` map.

Note: we really should deep-copy the `Value` implementations before setting them up as flag values here, but that's another story that was laziness on my part in the `flag` source as well. Maybe I'll move the deep copier into its own package after I finish fixing the cycle-handling, etc. Then we can trivially deep-copy the field-value for this and the TextUnmarshaler implementations.

Reply:
Made `fieldVal.Addr()` the value for `flagValues` map and changed around how the supported types (like complex number) are handling pointers to make this work. But yes, definitely need to deep copy the field value with all the pointers